firedrake/constant.py

```diff
@@ -67,7 +67,7 @@ def __init__(self, value, domain=None):
             e = ufl.TensorElement("Real", domain, 0, shape=shape)
         super(Constant, self).__init__(e)
         self._ufl_element = self.element()
-        self._repr = 'Constant(%r)' % self._ufl_element
+        self._repr = 'Constant(%r, %r)' % (self._ufl_element, self.count())
```

Comment:
In caches.  UFL computes the hash of a Terminal by computing the hash of the repr (ugh).